### ISSUE TITLE
Opening PDF is hanging

### ISSUE THREAD

**Issue by brandoncc** [libvips/ruby-vips#309]
I reinstalled vips using homebrew the other day and now I am unable to open pdf files with ruby-vips. I also upgraded to Big Sur a couple of months ago.

In IRB, I get:

> Vips::Error (VipsOperation: class "pdfload" not found)

I have:

- Uninstalled vips
- Uninstalled poppler (I had two recent versions installed with homebrew)
- Reinstalled poppler
- Reinstalled vips (8.11.0 now)
- Updated ruby-vips

```bash
$ vips --vips-config
enable debug: no
enable deprecated library components: yes
enable modules: yes
enable docs with gtkdoc: no
gobject introspection:  yes
RAD load/save: yes
Analyze7 load/save: yes
PPM load/save: yes
GIF load:  yes
generate C++ docs: no
use fftw3 for FFT: yes
accelerate loops with orc: yes
ICC profile support with lcms: yes (lcms2)
zlib: yes
text rendering with pangocairo: yes
font file support with fontconfig: yes
EXIF metadata support with libexif: yes
JPEG load/save with libjpeg: yes (pkg-config)
JXL load/save with libjxl: no (dynamic module: no)
JPEG2000 load/save with libopenjp2: yes
PNG load with libspng: yes
PNG load/save with libpng: yes (pkg-config libpng >= 1.2.9)
8bpp PNG quantisation: yes
TIFF load/save with libtiff: yes (pkg-config libtiff-4)
image pyramid save: yes
HEIC/AVIF load/save with libheif: yes (dynamic module: yes)
WebP load/save with libwebp: yes
PDF load with PDFium:  no
PDF load with poppler-glib: yes (dynamic module: yes)
SVG load with librsvg-2.0: yes
EXR load with OpenEXR: yes
slide load with OpenSlide: yes (dynamic module: yes)
Matlab load with matio: yes
NIfTI load/save with niftiio: no
FITS load/save with cfitsio: yes
Magick package: MagickCore (dynamic module: no)
Magick API version: magick7
load with libMagickCore: yes
save with libMagickCore: yes
```

I am able to open a pdf at the command line:

```bash
$  vips pdfload test.pdf test.jpg
$
```

And `test.jpg` is created.

In IRB:

```
irb(main):001:0> require 'vips'
=> true
irb(main):002:0> Vips::Image.pdfload('/Users/brandoncc/Downloads/test.pdf')
Traceback (most recent call last):
        9: from /Users/brandoncc/.asdf/installs/ruby/2.5.7/bin/irb:11:in `<main>'
        8: from (irb):2
        7: from /Users/brandoncc/.asdf/installs/ruby/2.5.7/lib/ruby/gems/2.5.0/gems/ruby-vips-2.1.2/lib/vips/image.rb:234:in `method_missing'
        6: from /Users/brandoncc/.asdf/installs/ruby/2.5.7/lib/ruby/gems/2.5.0/gems/ruby-vips-2.1.2/lib/vips/operation.rb:350:in `call'
        5: from /Users/brandoncc/.asdf/installs/ruby/2.5.7/lib/ruby/gems/2.5.0/gems/ruby-vips-2.1.2/lib/vips/operation.rb:170:in `get'
        4: from /Users/brandoncc/.asdf/installs/ruby/2.5.7/lib/ruby/gems/2.5.0/gems/ruby-vips-2.1.2/lib/vips/operation.rb:170:in `new'
        3: from /Users/brandoncc/.asdf/installs/ruby/2.5.7/lib/ruby/gems/2.5.0/gems/ruby-vips-2.1.2/lib/vips/operation.rb:67:in `initialize'
        2: from /Users/brandoncc/.asdf/installs/ruby/2.5.7/lib/ruby/gems/2.5.0/gems/ruby-vips-2.1.2/lib/vips/operation.rb:67:in `new'
        1: from /Users/brandoncc/.asdf/installs/ruby/2.5.7/lib/ruby/gems/2.5.0/gems/ruby-vips-2.1.2/lib/vips/operation.rb:204:in `initialize'
Vips::Error (VipsOperation: class "pdfload" not found)
```

In Rails Console:

```ruby
[1] pry(main)> Vips::Image.pdfload('/Users/brandoncc/Downloads/test.pdf')
_
```

It hangs in Ruby!

Any help would be greatly appreciated.

**Comment by jcupitt** [comment]
Hi @brandoncc,

It looks like you have a couple of libvipses installed. Your CLI is using an old version (with pdf support) and ruby is picking up the new homebrew one (pdf support seems to be missing). I'd remove the old libvips to reduce confusion.

Your second problem is that homebrew appear to have removed support for poppler. I'll have a look.

**Comment by jcupitt** [comment]
Ah it's the new dynamic module system -- it's not working on macOS. I'll dig a bit more.

**Comment by jcupitt** [comment]
libvips is unable to find the PDF load module. As a workaround you can do this:

```
$ export VIPSHOME=/usr/local
$ irb
irb(main):001:0> require "vips"
=> true
irb(main):002:0> Vips::Image.pdfload("something.pdf")
=> #<Image 303x496 uchar, 4 bands, srgb>
irb(main):003:0> 
```

I'll see if I can improve module discovery.

**Comment by jcupitt** [comment]
I've made some changes to the way libvips finds library files and it's working in homebrew now. These improvements will be in 8.11.1. In the meantime, I'd use VIPSHOME as a workaround.

**Comment by brandoncc** [comment]
Thank you @jcupitt !

**Comment by brandoncc** [comment]
Hello again @jcupitt,

IRB works now, but rails console is still hanging in an application where we have used ruby-vips for years. I haven't been able to figure out how to get a stack trace out of the hung process yet, but I will post it here if I manage to do so.

**Comment by brandoncc** [comment]
Fun....I just downgraded to 8.10.6 and it is still hanging 😦 

There goes the rest of my day debugging or just reinstalling my OS!

**Comment by brandoncc** [comment]
Okay, I got it working after downgrading. Things I did:

- `wget https://raw.githubusercontent.com/Homebrew/homebrew-core/7b09694a24d23218aa7228493d12f2ad56d0a8a6/Formula/vips.rb`
- `brew install vips.rb` (didn't work)
- `brew install -s vips.rb` (required me to upgrade xcode command line tools first) -- this worked
- `brew uninstall vips`
- `brew install vips` -- trying 8.11.0 again since command line tools were upgraded and I saw 8.10.6 start working again -- didn't work
- `brew uninstall vips`
- `brew install vips.rb` -- worked this time

I am now able to open pdfs again:

```ruby
[1] pry(main)> Vips::Image.pdfload('/Users/brandoncc/Downloads/test.pdf')
=> #<Image 594x594 uchar, 4 bands, srgb>
```

**Comment by jcupitt** [comment]
Another workaround would be to configure vips with `--disable-modules`. That'll stop it building the pdf loader as a dynamic module and instead build it as part of the main vips library.

**Comment by brandoncc** [comment]
Thanks, I'll keep an eye out for 8.11.1 and test that. If it doesn't resolve my issue, then I will go that route. I assume I must be able to pass that flag to homebrew's build somehow.

**Comment by jcupitt** [comment]
You can download the head of the 8.11 branch to try it out:

https://github.com/libvips/libvips/archive/refs/heads/8.11.zip

I think it should fix this issue (hopefully).

You need to edit your vips.rb to pass the flag. Change the line:

```ruby
    system "./configure", *args
```

To be:

```ruby
    system "./configure", "--disable-modules", *args
```

That ought to fix 8.11.0.
